pageserver/src/tenant/blob_io.rs

```diff
@@ -83,41 +84,101 @@ impl<'a> BlockCursor<'a> {
-    /// you write something directly to the inner Write object, it makes the
-    /// internally tracked 'offset' to go out of sync. So don't do that.
-    pub fn into_inner(self) -> W {
-        self.inner
+    const CAPACITY: usize = if BUFFERED { PAGE_SZ } else { 0 };
+
+    #[inline(always)]
+    /// Writes the given buffer directly to the underlying `VirtualFile`.
+    /// You need to make sure that the internal buffer is empty, otherwise
+    /// data will be written in wrong order.
+    async fn write_all_unbuffered(&mut self, src_buf: &[u8]) -> Result<(), Error> {
+        self.inner.write_all(src_buf).await?;
+        self.offset += src_buf.len() as u64;
+        Ok(())
+    }
+
+    #[inline(always)]
+    /// Flushes the internal buffer to the underlying `VirtualFile`.
+    pub async fn flush_buffer(&mut self) -> Result<(), Error> {
+        self.inner.write_all(&self.buf).await?;
+        self.buf.clear();
+        Ok(())
+    }
+
+    #[inline(always)]
+    /// Writes as much of `src_buf` into the internal buffer as it fits
+    fn write_into_buffer(&mut self, src_buf: &[u8]) -> usize {
+        let remaining = Self::CAPACITY - self.buf.len();
+        let to_copy = src_buf.len().min(remaining);
+        self.buf.extend_from_slice(&src_buf[..to_copy]);
+        self.offset += to_copy as u64;
+        to_copy
+    }
+
+    /// Internal, possibly buffered, write function
+    async fn write_all(&mut self, mut src_buf: &[u8]) -> Result<(), Error> {
+        if !BUFFERED {
+            if !self.buf.is_empty() {
+                // Flush the buffer. This creates a write call for
+                // potentially very small data, but there is no way
+                // we can unify it with the data we are writing below
+                // without copying it.
+                self.flush_buffer().await?;
+            }
+            self.write_all_unbuffered(src_buf).await?;
+            return Ok(());
+        }
+        let remaining = Self::CAPACITY - self.buf.len();
+        // First try to copy as much as we can into the buffer
+        if remaining > 0 {
+            let copied = self.write_into_buffer(src_buf);
+            src_buf = &src_buf[copied..];
+        }
+        // Then, if the buffer is full, flush it out
+        if self.buf.len() == Self::CAPACITY {
+            self.flush_buffer().await?;
+        }
+        // Finally, write the tail of src_buf:
+        // If it wholly fits into the buffer without
+        // completely filling it, then put it there.
+        // If not, write it out directly.
+        if !src_buf.is_empty() {
+            assert_eq!(self.buf.len(), 0);
+            if src_buf.len() < Self::CAPACITY {
+                let copied = self.write_into_buffer(src_buf);
+                // We just verified above that src_buf fits into our internal buffer.
+                assert_eq!(copied, src_buf.len());
+            } else {
+                self.write_all_unbuffered(src_buf).await?;
```

Comment:
I don't think we need to worry about this right now, but just as a note, this code path could be more efficient by issuing this write in parallel with the write from flush_buffer on line 163.

In future if/when we start caring about pipelining writes for max throughput, it'll also need to be done at a higher level in the layer classes.

Reply:
For example, currently in the case of a request with more bytes than fit in the buffer, we're doing:
1. copy the first bytes of the request into the end of the buffer
2. issue a write for the buffer
3. issue a write for the remainder of the request data

Steps 2 + 3 can in principle run concurrently, e.g. by creating two futures and doing a tokio::select on them.  This only works if:
- one has already extended the file far enough (e.g. with falloc), so that the write_at will work as expected if step 3 hit the filesystem before step 3.
- _and_ there are proper mechanisms in place to handle the crash case, since if we crash when step 3 is complete but step 2 is not, then there would be a gap in the file.

It's not a defect in your PR though: the existing code wasn't doing anything like this either.